PyPy blog (#20)
alloftech/alloftech.github.io: _posts/2022-10-08-02-pypy.md

---
layout: post
title: "Run python faster using PyPy"
date: 2022-10-08
last_modified_at: 2022-10-08
cover_image: 2022-10-08-00-pypy-logo.jpg
author: Sanket Benke
categories: Python
excerpt_separator: <!--more-->
---

>               "If you want your code to run faster,
                  you should probably just use PyPy."
                                            - Guido van Rossum (Creater of python)

> 

Reviewer: ```suggestion

> Photo by pypy(https://www.pypy.org/)

With support of huge number of libraries which enables provision for easy compatibility for any tool, **_Python_** has been one of the most used programming language for scripting. In this blog, we will look at how can we can easily increase the efficiency of our scripts multifold based on our usecase.

>               "If you want your code to run faster,
>                  you should probably just use PyPy."
>                                            - Guido van Rossum (Creater of python)

```

<!--more-->
<br><br>

> Lets first understand how python program is executed.

Reviewer: ```suggestion
<br><br>

### <span style="text-decoration: underline"> Default Python Execution </span>

Lets first understand how python program is executed.
```

|![CPython Implementation](/assets/images/2022-10-08-00-CPython-Implementation.png) |
|:--:|
| *CPython Implementation* |

 - First you write your python code in file, lets call that file foo.py.
 - You run python below command
   > python foo.py

 - Of course computer doesn't understand this high level language. Computer works only on machine languages. To convert your code to machine language, python perform following steps:

    - Python code gets compiled to bytecode and gives .pyc file, in our case foo.pyc
    - This .pyc file then gets interpreted to machine language in PVM(Python Virtual Machine) and finally you get your output

 - This process of implementation is **CPython**.

>  - **Python is a programming language and CPython is implementation of python.**

Reviewer: ```suggestion
- First you write your python code in file, lets call that file _foo.py_.
 - You execute your code using below command <br>
   `python foo.py`

 - Ofcourse, computer doesn't understand this high level language as computer works only on machine language. To convert your code to machine language, python performs following steps:
    - Python code gets compiled to **bytecode** and gives `.pyc` file, in our case _foo.pyc_
    - This `.pyc` file then gets **interpreted** to machine language in **PVM**(_Python Virtual Machine_) and finally you get your output
 - This process of implementation is referred as **CPython**.
 - **Python is a programming language & CPython is default implementation of python.**
```

<br><br>
 ## CPython:

 - CPython is the default byte-code interpreter of Python, which is written in C.
 - When you install python from https://www.python.org/, default iterpreter is CPython.
 - **CPython basically convert your code to bytecode first and then interpreted this bytecode to machine language and gives you output.**
 - CPython is slow

> <br><br>

Reviewer: ```suggestion

 ### <span style="text-decoration: underline"> Default Python (CPython) VS PyPy </span>

 ## CPython:

 - CPython is the default byte-code interpreter of Python, which is written in **C** programming language.
 - When you install python from the [official website](https://www.python.org/), default interpreter is **CPython**.
 - **CPython basically converts your code to bytecode first and then interprets this bytecode to machine language and gives you desired output.**
 - CPython tends to be slower due to multilevel conversion of your code.

<br>

```

 ## PyPy:

 - As we now know, CPython is slow, to fill this gap, PyPy is used.
 - PyPy is written in RPython(Restricted Python - a subset of python.)
 - But what makes PyPy fast? It is JIT Compiler.
 - **PyPy uses JIT or Just-in-Time compilation approach to compile the source code into native machine code instead of bytecode. Thereby making the code to run faster.**
 - PyPy is a compiler and CPython is interpreter.
 - PyPy workd faster for only pure python code.
 - So the case where PyPy works best is when executing long-running programs where a significant fraction of the time is spent executing Python code.

 - Features of PyPy:
   - Speed (4.4 times faster than CPython on an average.)
   - Memory usage
   - Stackless

 - Drawbacks of PyPy:
   - if your code doesn't run for at least a few seconds, then the JIT compiler won't have enough time to warm up.
   - If all the time is spent in run-time libraries (i.e. in C functions), and not actually running Python code, the JIT compiler will not help.
   - Not suitable for data science libraries as it contains C functions and not pure python code.

> <br><br>

Reviewer: ```suggestion

 ## PyPy:

 - As we now understand the reason for CPython to be slow, to fill this gap, **PyPy** is used.
 - PyPy is written in **RPython**(_Restricted Python - a subset of python_)
 - But what makes PyPy faster ? It is **JIT Compiler**.
 - **PyPy uses JIT or Just-in-Time compilation approach to compile the source code directly into native machine code instead of bytecode. Thereby making the code to run faster by skipping any intermediate conversion.**
 - PyPy is a compiler while CPython is an interpreter.
 - PyPy works faster for only **pure python code** i.e. there shouldn't be any embedded C code. If that's the case, it's performance deteriorates. 
 - So the use case where PyPy works best is when executing long-running programs where a significant fraction of the time is spent executing Python code.

 - Features of PyPy:
   - Significant increase in execution speed (4.4 times faster than CPython on an average.)
   - Efficient memory usage
   - Stackless

 - Drawbacks of PyPy:
   - if your code doesn't run for at least a few seconds, then the JIT compiler won't have enough time to warm up.
   - If all the time is spent in run-time libraries (i.e. in C functions), and not actually running Python code, the JIT compiler will not be any better than usual execution.
   - Not suitable for data science libraries as those contain C functions and not pure python code.
   
   <br><br>
```

## Installation of PyPy3:

- Ubuntu:

  > sudo add-apt-repository ppa:pypy/ppa

  > sudo apt update

  > sudo apt install pypy3

- Mac:

  > brew install pypy3

- Windows:

>   > Download Windows Binary file from https://www.pypy.org/download.html

Reviewer: ```suggestion
### <span style="text-decoration: underline"> Installion and setting up of PyPy3: </span>

Let's walk through installation of PyPy3 on different machines.
<br>

- Ubuntu:

  - Adding pypy to repository<br>
    `sudo add-apt-repository ppa:pypy/ppa`

  - Updating packages<br>
    `sudo apt update`

  - Installing pypy3<br>
    `sudo apt install pypy3`

- Mac:

  - Installing pypy3 using _brew_ <br> 
    `brew install pypy3`

- Windows:

  - Download Windows Binary file from [official website](https://www.pypy.org/download.html)
  - Execute the binary for installation
```

<br>
## Installation of Pip:

 - Download the latest version of get-pip.py from https://bootstrap.pypa.io/.
 - go to the folder where you download get-pip and run following cmd:
   > pypy get-pip.py

## Running python code using pypy:

 - First write your python code.
 - then run following command instead of running > python3 <file_name.py>

>    > pypy3 <file_name.py>

Reviewer: ```suggestion
Along with PyPy, we would also need _pip_ for python package management

- Installation of Pip:
  - Download the latest version of [get-pip.py](https://bootstrap.pypa.io/.)
  - Go to the folder where you download get-pip and run following cmd:
    `pypy get-pip.py`

Now finally, it's similar to use PyPy just like using Cpython for python code execution

- Running python code using pypy:
  - First write your python code.
  - then run following command instead of running `python3 <file_name.py>`
    `pypy3 <file_name.py>`
```

<br><br>

> ## Comparison between CPython and PyPy
> 
>  |![CPython](/assets/images/2022-10-08-00-python3-time-taken.png) |
>  |:--:|
>  | *time taken by CPython* |
> 
> 
>  |![PyPy](/assets/images/2022-10-08-00-pypy3-time-taken.png) |
>  |:--:|
>  | *time taken by PyPy* |
> 
> 
> As we can see by above example, PyPy took only 5 seconds to complete and CPython took 2 minutes and 20 seconds to complete.
> So use PyPy for pure python code and use CPython if you have libraries which are in C and python language.
> 
> 
> References:
> - https://www.pypy.org/

Reviewer: ```suggestion
### <span style="text-decoration: underline"> CPython & PyPy Execution </span>

Let's check upon actual execution time of same piece of code in both Cpython & PyPy for getting an idea of its difference! <br>
For comparision, we will execute a long iterating `for` loop and record execution period using `time`.

 |![CPython](/assets/images/2022-10-08-00-python3-time-taken.png) |
 |:--:|
 | *time taken by CPython* |

 |![PyPy](/assets/images/2022-10-08-00-pypy3-time-taken.png) |
 |:--:|
 | *time taken by PyPy* |

 > As we can see by above example, PyPy took only **5 seconds** to complete while CPython took **2 minutes and 20 seconds** to complete execution of same loop.

### <span style="text-decoration: underline"> Conclusion </span>

No doubt, Python (CPython) is vastly used for ML/AI projects as most of its libraries have embedded C functions included. So in such case, PyPy won't serve its purpose well !<br>
But for usecase involving pure python code, it will be best to use [**PyPy**](https://www.pypy.org/) instead of **CPython** for better execution efficiency. <br> 

```
